File: Demo/BSTextDemo/Examples/CTDrawView.m
//
//  CTDrawView.m
//  BSTextDemo
//
//  Created by jeremy on 2024/4/7.
//  Copyright © 2024 GeekBruce. All rights reserved.
//

#import "CTDrawView.h"
#import <CoreText/CoreText.h>

typedef enum : NSUInteger {
    YTDrawModeLines,
    YTDrawModeFrame,
} YTDrawMode;

@interface YTCTLine : NSObject

@property (nonatomic, assign) CGPoint position;
@property (nonatomic, assign) CTLineRef ctLine;

@end

@implementation YTCTLine

- (void)setCtLine:(CTLineRef)ctLine{
  
  if (_ctLine) {
    CFRelease(_ctLine);
  }
  if (ctLine) {
    CFRetain(ctLine);
  }
  _ctLine = ctLine;
}

@end

@interface CTDrawView()

@property(nonatomic, strong)NSMutableArray<YTCTLine *> *linesToDraw;


/**
 绘制模式，使用CTFrame或者CTLine
 */
@property (nonatomic, assign) YTDrawMode drawMode;

@property (nonatomic, assign) CTFrameRef ctFrame;

@property(nonatomic, strong)NSAttributedString *attributeString;

@property (nonatomic, strong) NSAttributedString *truncationToken;///<截断的标识字符串，默认是"..."


@end

@implementation CTDrawView

- (instancetype)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    self.textColor = [UIColor redColor];
  }
  return self;
}

- (void)setText:(NSString*)text{
  [self.linesToDraw removeAllObjects];
  self.attributeString = [[NSAttributedString alloc]initWithString:text attributes:[self getStirngAttributes]];
}

- (void)setAttributeString:(NSAttributedString *)attributeString{
  _attributeString = attributeString;
  [self calculateTruncatedLinesWithBounds:self.bounds];
  [self setNeedsDisplay];
}

- (void)setCtFrame:(CTFrameRef)ctFrame{
  if(_ctFrame != nil){
    CFRelease(_ctFrame);
  }
  _ctFrame = ctFrame;
}

- (CTFrameRef)composeCTFrameWithAttributeString:(NSAttributedString *)attributeString frame:(CGRect)frame {
   
    // 使用NSMutableAttributedString创建CTFrame
    CTFramesetterRef ctFramesetter = CTFramesetterCreateWithAttributedString((CFAttributedStringRef)attributeString);
    CGSize size =  CTFramesetterSuggestFrameSizeWithConstraints(ctFramesetter, CFRangeMake(0, attributeString.length), NULL, CGSizeMake(frame.size.width, CGFLOAT_MAX), NULL);
  
    // 绘制区域
    CGMutablePathRef path = CGPathCreateMutable();
    CGPathAddRect(path, NULL, CGRectMake(frame.size.width/2.0 - size.width/2.0,frame.size.height-size.height, size.width, size.height));

    CTFrameRef ctFrame = CTFramesetterCreateFrame(ctFramesetter, CFRangeMake(0, attributeString.length), path, NULL);
    
    CFRelease(ctFramesetter);
    CFRelease(path);
    
    return ctFrame;
}

- (void)drawRect:(CGRect)rect {
    [super drawRect:rect];

    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetTextMatrix(context, CGAffineTransformIdentity);
    CGContextTranslateCTM(context, 0, self.bounds.size.height);
    CGContextScaleCTM(context, 1, -1);
    
    // 绘制文字
    [self drawTextInContext:context];
}



    - (void)calculateTruncatedLinesWithBounds:(CGRect)bounds {
      
      self.ctFrame = [self composeCTFrameWithAttributeString:self.attributeString frame:self.bounds];
      
      NSArray *lines = (NSArray *)CTFrameGetLines(self.ctFrame);
      NSUInteger lineCount = lines.count;
      
      if (self.numberOfLines >0 && lineCount>self.numberOfLines) {
        // need truncate
        self.drawMode = YTDrawModeLines;
        CGPoint lineOrigins[lineCount];
        CTFrameGetLineOrigins(self.ctFrame, CFRangeMake(0,lineCount), lineOrigins);
        for (NSUInteger i = lineCount-2; i<lineCount; i++) {
          CTLineRef line = (__bridge CTLineRef)(lines[i]);
          CGFloat ascent, descent;
          CGFloat lineWidth = CTLineGetTypographicBounds(line, &ascent, &descent, NULL);
          if (i == lineCount-2) {
            //next-to-last line
            NSAttributedString* tokenString = [[NSAttributedString alloc] initWithString:@"\u2026" attributes:[self getStirngAttributes]];
            CTLineRef truncationTokenLine = CTLineCreateWithAttributedString((__bridge CFAttributedStringRef)tokenString);
            CGFloat truncationTokenLineWidth = CTLineGetTypographicBounds(truncationTokenLine, NULL, NULL, NULL);
            //generate new lines
            CTLineRef lastLine = CTLineCreateTruncatedLine(line,lineWidth-truncationTokenLineWidth, kCTLineTruncationStart, truncationTokenLine);
            lineWidth = CTLineGetTypographicBounds(lastLine, &ascent, &descent,NULL);
            YTCTLine *ytLine = [YTCTLine new];
            ytLine.ctLine = lastLine;
            ytLine.position = CGPointMake(bounds.size.width/2-lineWidth/2,self.bounds.size.height-ascent);
            [self.linesToDraw addObject:ytLine];
            
          }else{
            //last line
            YTCTLine* ytline = [YTCTLine new];
            ytline.ctLine = line;
            ytline.position =  CGPointMake(bounds.size.width/2- lineWidth/2,self.bounds.size.height - (ascent+descent)-ascent);
            [self.linesToDraw addObject:ytline];
          }
        }
      }else{
        self.drawMode = YTDrawModeFrame;
      }
    }


- (NSAttributedString *)attributeStringToDraw {
    return self.attributeString;
}
  
  
  - (NSDictionary*)getStirngAttributes{
    NSMutableParagraphStyle* paraghStyle = [[NSMutableParagraphStyle alloc]init];
    paraghStyle.alignment = NSTextAlignmentCenter;
    NSDictionary *tokenAttributes =@{NSForegroundColorAttributeName:self.textColor,NSFontAttributeName:[UIFont fontWithName:@"PingFangSC-Regular" size:22],NSParagraphStyleAttributeName:paraghStyle};
    return  tokenAttributes;
}

/**
 绘制文字
 */
- (void)drawTextInContext:(CGContextRef)context {
    if (self.drawMode == YTDrawModeFrame) {
      CTFrameDraw(self.ctFrame, context);
    } else if (self.drawMode == YTDrawModeLines) {
        for (YTCTLine *line in self.linesToDraw) {
            // 设置Line绘制的位置
            CGContextSetTextPosition(context, line.position.x, line.position.y);
            CTLineDraw(line.ctLine, context);
        }
    }
}





- (NSMutableArray<YTCTLine *> *)linesToDraw {
    if (!_linesToDraw) {
        _linesToDraw = [NSMutableArray array];
    }
    return _linesToDraw;
}


@end

File: Demo/BSTextDemo/Examples/CTDrawView.swift
//
//  LMDrawView.swift
//  BSTextDemo
//
//  Created by jeremy on 2024/4/6.
//  Copyright © 2024 GeekBruce. All rights reserved.
//

import UIKit
import CoreText

class YTCTLine:NSObject{
  var position:CGPoint
  var ctLine:CTLine
  
  
  init(position: CGPoint, ctLine: CTLine) {
    self.position = position
    self.ctLine = ctLine
  }
}

public class CTDrawView:UIView{
  
  var numberOfLines:Int = 2
  
  var ctframe:CTFrame? = nil
  
  var lineArray:[YTCTLine] = []
  var attriString:NSAttributedString? {
    didSet{
      guard let attriString = attriString else {
        return
      }
      let frameSetter = CTFramesetterCreateWithAttributedString(attriString)
      var size = CGSize(width:self.bounds.size.width, height: CGFloat(MAXFLOAT))
      size =  CTFramesetterSuggestFrameSizeWithConstraints(frameSetter, CFRange(location: 0, length: 0), nil, size, nil)
      let path = CGPath(rect: CGRect(origin: CGPoint(x: 0, y: 0), size: size), transform: nil)
      let frame = CTFramesetterCreateFrame(frameSetter,CFRange(location: 0, length: attriString.length), path, nil)
      
      var lines = CTFrameGetLines(frame)
      var lineCount = CFArrayGetCount(lines)
      var lineOriginArray = UnsafeMutablePointer<CGPoint>.allocate(capacity: lineCount)
      CTFrameGetLineOrigins(frame, CFRange(location: 0, length: lineCount), lineOriginArray)
      
      guard self.numberOfLines != 0 && lineCount > self.numberOfLines else{
        self.ctframe = frame
        return
      }
      
      for i in (0..<self.numberOfLines).reversed(){
        let ctLine = unsafeBitCast(CFArrayGetValueAtIndex(lines, i), to: CTLine.self)
        let runs = CTLineGetGlyphRuns(ctLine)
        if CFArrayGetCount(runs) == 0{
          continue
        }
        let range = CTLineGetStringRange(ctLine)
        if i == 0 {
         let ytctLine = YTCTLine.init(position: CGPoint(x: 0, y: 0), ctLine: ctLine)
          self.lineArray.insert(ytctLine, at: 0)
        }else{
          let tokenString = NSAttributedString(string: "\u2026",attributes:nil)
          let tokenSize = tokenString.boundingRect(with: CGSize(width: CFloat(MAXFLOAT), height: CFloat(MAXFLOAT)), options: .usesLineFragmentOrigin, context: nil).size
          
          
        }
      }

      
      
    }
  }
  
  public override func draw(_ rect: CGRect) {
    
  }
  
}
